File: android/app/src/main/java/com/ivanlee/financetracker/logic/CategoryPeriod.kt
package com.ivanlee.financetracker.logic

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset

/**
 * The date window the Activity screen's "Top Categories" card is scoped to.
 *
 * Kotlin port of the web's `CategoryPeriodPreset`
 * (`frontend/src/pages/Transactions/Transactions.tsx`) and iOS's `Support/CategoryPeriod.swift`.
 * [wire] matches the web's stored strings so all three clients name the same window the same way.
 */
enum class CategoryPeriod(val wire: String, val label: String) {
    ALL("all", "All time"),
    THIS_MONTH("this_month", "This month"),
    LAST_MONTH("last_month", "Last month"),
    LAST_3_MONTHS("last_3_months", "Last 3 months"),
    LAST_6_MONTHS("last_6_months", "Last 6 months"),
    THIS_YEAR("this_year", "This year"),
    SPECIFIC_MONTH("specific_month", "Specific month"),
    CUSTOM("custom", "Custom range"),
    ;

    /** Whether this case needs the custom start/end dates filled in to mean anything. */
    val usesCustomStart: Boolean get() = this == CUSTOM || this == SPECIFIC_MONTH

    companion object {
        fun fromWire(value: String?): CategoryPeriod =
            entries.firstOrNull { it.wire == value } ?: ALL
    }
}

/**
 * A half-open window: [start] inclusive, [end] **exclusive**. Either side may be null, meaning
 * unbounded in that direction.
 */
data class CategoryPeriodRange(val start: Instant?, val end: Instant?) {
    operator fun contains(date: Instant): Boolean {
        if (start != null && date < start) return false
        if (end != null && date >= end) return false
        return true
    }
}

/**
 * The window to filter by, or null for "no filter at all" ([CategoryPeriod.ALL], or a custom range
 * with neither end filled in).
 *
 * Boundaries are computed in **UTC**, the same rule the rest of this client follows: backend
 * transaction dates are naive and read back as UTC, so a local-calendar month boundary would push
 * the 1st of the month into the previous bucket for anyone west of Greenwich.
 */
fun CategoryPeriod.range(
    customStart: Instant? = null,
    customEnd: Instant? = null,
    now: Instant = Instant.now(),
): CategoryPeriodRange? {
    val today = now.atZone(ZoneOffset.UTC).toLocalDate()
    fun monthStart(offset: Long): Instant =
        today.withDayOfMonth(1).plusMonths(offset).atStartOfDay(ZoneOffset.UTC).toInstant()

    fun startOfDay(instant: Instant): Instant =
        instant.atZone(ZoneOffset.UTC).toLocalDate().atStartOfDay(ZoneOffset.UTC).toInstant()

    return when (this) {
        CategoryPeriod.ALL -> null
        CategoryPeriod.THIS_MONTH -> CategoryPeriodRange(monthStart(0), null)
        CategoryPeriod.LAST_MONTH -> CategoryPeriodRange(monthStart(-1), monthStart(0))
        CategoryPeriod.LAST_3_MONTHS -> CategoryPeriodRange(monthStart(-2), null)
        CategoryPeriod.LAST_6_MONTHS -> CategoryPeriodRange(monthStart(-5), null)
        CategoryPeriod.THIS_YEAR -> CategoryPeriodRange(
            LocalDate.of(today.year, 1, 1).atStartOfDay(ZoneOffset.UTC).toInstant(),
            null,
        )
        CategoryPeriod.SPECIFIC_MONTH -> {
            // `customStart` doubles as the month anchor — only its year/month are read, so any
            // day within the month selects that whole month. Reusing the field keeps the saved
            // preferences one shape rather than adding a third date that's null most of the time.
            if (customStart == null) {
                null
            } else {
                val anchor = customStart.atZone(ZoneOffset.UTC).toLocalDate().withDayOfMonth(1)
                CategoryPeriodRange(
                    anchor.atStartOfDay(ZoneOffset.UTC).toInstant(),
                    anchor.plusMonths(1).atStartOfDay(ZoneOffset.UTC).toInstant(),
                )
            }
        }
        CategoryPeriod.CUSTOM -> {
            val start = customStart?.let { startOfDay(it) }
            // The end date the user picked is inclusive, so the exclusive bound is the next midnight.
            val end = customEnd?.let { startOfDay(it).plus(java.time.Duration.ofDays(1)) }
            if (start == null && end == null) null else CategoryPeriodRange(start, end)
        }
    }
}

/**
 * The user's saved Top-Categories filter for one household: which categories are hidden and which
 * period is selected. Persisted so it survives leaving the tab, matching the web's localStorage.
 */
data class TopCategoryFilterPrefs(
    val hiddenCategoryIds: Set<String> = emptySet(),
    val period: CategoryPeriod = CategoryPeriod.ALL,
    val customStart: Instant? = null,
    val customEnd: Instant? = null,
)

File: ios/FinanceTracker/Support/CategoryPeriod.swift
import Foundation

/// The date window the Transactions "Top categories" card is scoped to.
///
/// Swift port of the web's `CategoryPeriodPreset` (`frontend/src/pages/Transactions/Transactions.tsx`);
/// the Kotlin twin is `android/.../logic/CategoryPeriod.kt`. Raw values match the web's strings so the
/// three clients describe the same window with the same name.
enum CategoryPeriod: String, CaseIterable, Identifiable, Codable {
    case all
    case thisMonth = "this_month"
    case lastMonth = "last_month"
    case last3Months = "last_3_months"
    case last6Months = "last_6_months"
    case thisYear = "this_year"
    case specificMonth = "specific_month"
    case custom

    var id: String { rawValue }

    var label: String {
        switch self {
        case .all: "All time"
        case .thisMonth: "This month"
        case .lastMonth: "Last month"
        case .last3Months: "Last 3 months"
        case .last6Months: "Last 6 months"
        case .thisYear: "This year"
        case .specificMonth: "Specific month"
        case .custom: "Custom range"
        }
    }

    /// Whether this case needs the custom start/end dates filled in to mean anything.
    var usesCustomStart: Bool { self == .custom || self == .specificMonth }
}

/// A half-open window: `start` inclusive, `end` **exclusive**. Either side may be nil, meaning
/// unbounded in that direction.
struct CategoryPeriodRange: Equatable {
    let start: Date?
    let end: Date?

    func contains(_ date: Date) -> Bool {
        if let start, date < start { return false }
        if let end, date >= end { return false }
        return true
    }
}

extension CategoryPeriod {
    /// Boundaries are computed in **UTC**, not the device's calendar. Backend transaction dates are
    /// naive and read back as UTC, so a local-calendar month boundary would pull the 1st of the month
    /// into the previous bucket for anyone west of Greenwich — the same reason the Activity list groups
    /// its months in UTC.
    static var utcCalendar: Calendar {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        return cal
    }

    /// The window to filter by, or `nil` for "no filter at all" (`.all`, or a `.custom` range with
    /// neither end filled in).
    func range(customStart: Date? = nil, customEnd: Date? = nil, now: Date = Date()) -> CategoryPeriodRange? {
        let cal = Self.utcCalendar

        func monthStart(offset: Int) -> Date {
            let current = cal.date(from: cal.dateComponents([.year, .month], from: now))!
            return cal.date(byAdding: .month, value: offset, to: current)!
        }

        switch self {
        case .all:
            return nil
        case .thisMonth:
            return CategoryPeriodRange(start: monthStart(offset: 0), end: nil)
        case .lastMonth:
            return CategoryPeriodRange(start: monthStart(offset: -1), end: monthStart(offset: 0))
        case .last3Months:
            return CategoryPeriodRange(start: monthStart(offset: -2), end: nil)
        case .last6Months:
            return CategoryPeriodRange(start: monthStart(offset: -5), end: nil)
        case .thisYear:
            return CategoryPeriodRange(start: cal.date(from: cal.dateComponents([.year], from: now))!, end: nil)
        case .specificMonth:
            // `customStart` doubles as the month anchor — only its year/month are read, so any
            // day within the month selects that whole month. Reusing the field keeps the saved
            // preferences one shape rather than adding a third date that's null most of the time.
            guard let anchor = customStart else { return nil }
            let start = cal.date(from: cal.dateComponents([.year, .month], from: anchor))!
            return CategoryPeriodRange(start: start, end: cal.date(byAdding: .month, value: 1, to: start)!)
        case .custom:
            // The end date the user picked is inclusive, so the exclusive bound is the next midnight.
            let start = customStart.map { cal.startOfDay(for: $0) }
            let end = customEnd.map { cal.date(byAdding: .day, value: 1, to: cal.startOfDay(for: $0))! }
            if start == nil && end == nil { return nil }
            return CategoryPeriodRange(start: start, end: end)
        }
    }
}

/// The user's saved Top-categories filter for one household: which categories are hidden and which
/// period is selected. Persisted so it survives leaving the tab, matching the web's localStorage.
struct TopCategoryFilterPrefs: Codable, Equatable {
    var hiddenCategoryIds: Set<String> = []
    var period: CategoryPeriod = .all
    var customStart: Date?
    var customEnd: Date?

    static let empty = TopCategoryFilterPrefs()
}

/// UserDefaults-backed store for `TopCategoryFilterPrefs`, keyed per household — two households have
/// different categories, so one household's hidden set is meaningless in the other.
enum TopCategoryFilterStore {
    static func key(householdId: String) -> String { "ft:tx-category-filter:\(householdId)" }

    static func load(householdId: String, defaults: UserDefaults = .standard) -> TopCategoryFilterPrefs {
        guard let data = defaults.data(forKey: key(householdId: householdId)),
              let prefs = try? JSONDecoder().decode(TopCategoryFilterPrefs.self, from: data)
        else { return .empty }
        return prefs
    }

    static func save(_ prefs: TopCategoryFilterPrefs, householdId: String, defaults: UserDefaults = .standard) {
        guard let data = try? JSONEncoder().encode(prefs) else { return }
        defaults.set(data, forKey: key(householdId: householdId))
    }
}
